jinahub/rankers/DPRReaderRanker/dpr_reader.py

```diff
@@ -0,0 +1,203 @@
+        match's ``text`` attribute.
+
+        :param docs: Documents whose matches to re-rank (specifically, the matches of
+            the documents on the traversal paths will be re-ranked). The document's
+            ``text`` attribute is taken as the question, and the ``text`` attribute
+            of the matches as the context. The matches must also have a title, which
+            should be available in the tag with the key specified by the
+            ``title_tag_key`` parameter passed on the executor initialization.
+        :param parameters: dictionary to define the ``traversal_path`` and the
+            ``batch_size``. For example,
+            ``parameters={'traversal_paths': ['r'], 'batch_size': 10}``
+        """
+
+        if not docs:
+            return None
+
+        for doc in docs.traverse_flat(
+            parameters.get('traversal_paths', self.default_traversal_paths)
+        ):
+            if not doc.text:
+                raise ValueError(
+                    f'No question (text) found for document with id {doc.id}.'
+                    ' DPRReaderRanker requires a question from the main document'
+                    ' and context (text) from its matches.'
+                )
+
+            answer_spans = []
+            answer_relevance_scores = []
+            answer_span_scores = []
+            answer_titles = []
+
+            match_batches_generator = _batcher(
+                doc.matches, n=parameters.get('batch_size', self.default_batch_size)
+            )
+            for matches in match_batches_generator:
+                contexts = matches.get_attributes('text')
+                titles = matches.get_attributes(f'tags__{self.title_tag_key}')
+
+                if len(titles) != len(matches):
+                    raise ValueError(
+                        'All matches are required to have the'
+                        f' {self.title_tag_key} tag, but found'
+                        f' {len(titles) - len(matches)} matches without it.'
+                    )
+
+                with torch.no_grad():
+                    (
+                        relevance_scores,
+                        span_scores,
+                        spans,
+                        titles,
+                    ) = self._get_outputs(doc.text, contexts, titles)
+
+                answer_spans += spans
+                answer_relevance_scores += relevance_scores
+                answer_span_scores += span_scores
+                answer_titles += titles
+
+            # Make sure answers are sorted by relevance scores
+            sorting_list = list(zip(answer_relevance_scores, answer_span_scores))
+            answer_spans = _sort_by(answer_spans, sorting_list)
+
+            answer_titles = _sort_by(answer_titles, sorting_list)
+            answer_relevance_scores = _sort_by(answer_relevance_scores, sorting_list)
+            answer_span_scores = _sort_by(answer_span_scores, sorting_list)
```

Comment:
Even if it is needed, we are doing 4 sorts. We could sort only once and avoid zipping and only keep the valid indices

Reply:
Ok I'll replace it